File: tdf-hq-ui/src/api/pipelines.ts
import { del, get, patch, post } from './client';
import type {
  PipelineCardCreate,
  PipelineCardDTO,
  PipelineCardUpdate,
  PipelineDefinitionDTO,
  PipelineSnapshotDTO,
  PipelineStageDTO,
} from './types';

function requiredIdentifier(value: string, field: string): string {
  const normalized = value.trim();
  if (!normalized) throw new Error(`${field} is required.`);
  return encodeURIComponent(normalized);
}

export const Pipelines = {
  snapshot: () => get<PipelineSnapshotDTO>('/pipelines/snapshot'),
  definitions: () => get<PipelineDefinitionDTO[]>('/pipelines/definitions'),
  list: (workflowId: string) =>
    get<PipelineCardDTO[]>(`/pipelines/${requiredIdentifier(workflowId, 'workflowId')}`),
  stages: (workflowId: string) =>
    get<PipelineStageDTO[]>(`/pipelines/${requiredIdentifier(workflowId, 'workflowId')}/stages`),
  create: (workflowId: string, body: PipelineCardCreate) =>
    post<PipelineCardDTO>(`/pipelines/${requiredIdentifier(workflowId, 'workflowId')}`, body),
  get: (workflowId: string, cardId: string) =>
    get<PipelineCardDTO>(
      `/pipelines/${requiredIdentifier(workflowId, 'workflowId')}/${requiredIdentifier(cardId, 'cardId')}`,
    ),
  update: (workflowId: string, cardId: string, body: PipelineCardUpdate) =>
    patch<PipelineCardDTO>(
      `/pipelines/${requiredIdentifier(workflowId, 'workflowId')}/${requiredIdentifier(cardId, 'cardId')}`,
      body,
    ),
  remove: (workflowId: string, cardId: string) =>
    del<void>(
      `/pipelines/${requiredIdentifier(workflowId, 'workflowId')}/${requiredIdentifier(cardId, 'cardId')}`,
    ),
};

File: tdf-hq/src/TDF/API/Pipelines.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeOperators #-}

module TDF.API.Pipelines where

import           Data.Text (Text)
import           Servant

import           TDF.API.Types

type PipelinesAPI =
  "pipelines" :>
    (    "snapshot" :> Get '[JSON] PipelineSnapshotDTO
     :<|> "definitions" :> Get '[JSON] [PipelineDefinitionDTO]
     :<|> Capture "workflowId" Text :>
           (    Get '[JSON] [PipelineCardDTO]
            :<|> "stages" :> Get '[JSON] [PipelineStageDTO]
            :<|> ReqBody '[JSON] PipelineCardCreate :> PostCreated '[JSON] PipelineCardDTO
            :<|> Capture "cardId" Text :>
                  (    Get '[JSON] PipelineCardDTO
                   :<|> ReqBody '[JSON] PipelineCardUpdate :> Patch '[JSON] PipelineCardDTO
                   :<|> DeleteNoContent
                  )
           )
    )
